src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -181,6 +200,66 @@ private BoundExpression VisitCollectionBuilderCollectionExpression(BoundCollecti
                 type: constructMethod.ReturnType);
+
+            return new BoundSequence(
+                syntax,
+                locals.ToImmutableAndFree(),
+                sideEffects.ToImmutableAndFree(),
+                call,
+                call.Type);
+        }
+
+        private BoundExpression CreateAndPopulateInlineArray(
+            SyntaxNode syntax,
+            TypeWithAnnotations elementType,
+            ImmutableArray<BoundExpression> elements,
+            ArrayBuilder<LocalSymbol> locals,
+            ArrayBuilder<BoundExpression> sideEffects)
+        {
+            Debug.Assert(elements.Length > 0);
+            Debug.Assert(_factory.ModuleBuilderOpt is { });
+            Debug.Assert(_diagnostics.DiagnosticBag is { });
+            Debug.Assert(_compilation.Assembly.RuntimeSupportsInlineArrayTypes);
+
+            int arrayLength = elements.Length;
+            var inlineArrayType = _factory.ModuleBuilderOpt.EnsureInlineArrayTypeExists(syntax, _factory, arrayLength, _diagnostics.DiagnosticBag).Construct(ImmutableArray.Create(elementType));
+            Debug.Assert(inlineArrayType.HasInlineArrayAttribute(out int inlineArrayLength) && inlineArrayLength == arrayLength);
+
+            var intType = _factory.SpecialType(SpecialType.System_Int32);
+            var elementRef = _factory.ModuleBuilderOpt.EnsureInlineArrayElementRefExists(syntax, intType, _diagnostics.DiagnosticBag);
+            elementRef = elementRef.Construct(ImmutableArray.Create(TypeWithAnnotations.Create(inlineArrayType), elementType));
+
+            // Create an inline array and assign to a local.
+            BoundAssignmentOperator assignmentToTemp;
+            BoundLocal inlineArrayLocal = _factory.StoreToTemp(new BoundDefaultExpression(syntax, inlineArrayType), out assignmentToTemp, isKnownToReferToTempIfReferenceType: true);
+            sideEffects.Add(assignmentToTemp);
+            locals.Add(inlineArrayLocal.LocalSymbol);
+
+            for (int i = 0; i < arrayLength; i++)
+            {
+                var element = VisitExpression(elements[i]);
+                var call = _factory.Call(null, elementRef, inlineArrayLocal, _factory.Literal(i), useStrictArgumentRefKinds: true);
+                var assignment = new BoundAssignmentOperator(syntax, call, element, type: call.Type) { WasCompilerGenerated = true };
+                sideEffects.Add(assignment);
+            }
+
+            // Get a span to the inline array.
+            var spanType = _factory.WellKnownType(WellKnownType.System_Span_T); // PROTOTYPE: Test missing type.
+            var inlineArrayAsSpan = _factory.ModuleBuilderOpt.EnsureInlineArrayAsSpanExists(syntax, spanType, intType, _diagnostics.DiagnosticBag);
+            inlineArrayAsSpan = inlineArrayAsSpan.Construct(ImmutableArray.Create(TypeWithAnnotations.Create(inlineArrayType), elementType));
+
+            spanType = spanType.Construct(ImmutableArray.Create(elementType));
```

Comment:
This was already constructed by `EnsureInlineArrayAsSpanExists`/`SynthesizedInlineArrayAsSpanMethod`. Consider grabbing the return type from there.